### ISSUE TITLE
Ivy mode with 8.0.0-rc.2: Ivy's error message doesn't help to localize the error

### ISSUE THREAD

**Issue by juergenzimmermann** [angular/angular#30235]
# 🐞 bug report

### Affected Package
@angular/compiler

### Is this a regression?

I just tried to migrate to Ivy mode: edited tsconfig.app.json ("enableIvy": true, "allowEmptyCodegenFiles": true), and package.json ("postinstall": "ivy-ngcc"), and finally run "npm install".

### Description

The Ivy compiler issues an error message which doesn't help to localize the error:

## 🔥 Exception or Error

I get such an error message:

`ERROR in __ng_typecheck__.ts(108,41): error TS2532: Object is possibly 'undefined'.`

After painful stripping down the whole angular app and removing lots of code I found that Ivy has troubles with these lines:
```
<section class="mt-1" *ngIf="buch !== undefined">
   <hs-details-bearbeiten [id]="buch._id"></hs-details-bearbeiten>
```

## 🌍  Your Environment

Angular CLI 8.0.0-rc.2, Windows 10 Enterprise

**Angular Version:**

8.0.0-rc.2

**Comment by JoostK** [comment]
Thanks for reporting! The error you are seeing is emitted by Ivy's type checking mechanism, which currently does not provide information on which Angular template is responsible for the errors. This is being worked on in #30181.

In this particular case however, you have identified a false positive in the type-checker. As of today, decorators can specify a template guard function to narrow the type of a given expression. For `NgIf` this looks as follows:

```typescript
export declare class NgIf {
  static ngTemplateGuard_ngIf<E>(dir: NgIf, expr: E): expr is NonNullable<E>
}
```

For example, a template with `*ngIf="person"` with `person` being of type `Person | undefined`, the following code would be generated:

```typescript
if (NgIf.ngTemplateGuard_ngIf(directive, person) {
  // Usages of `person` within this block would be
  // of type `NonNullable<Person>`, which is just `Person`
}
```

In your case however, the expression passed to `ngIf`'s template guard is `buch !== undefined`:

```typescript
if (NgIf.ngTemplateGuard_ngIf(directive, buch !== undefined)) {
  // ...
}
```

This structure won't allow TypeScript to narrow the type of `buch` not to include `undefined`, so any usages of `buch` within the if-block are considered unsafe.

This is a current limitation of the generated type check code and we may be able to improve it by allowing the type of `NgIf.ngTemplateGuard_ngIf` to be something special that indicates to the type check generator that the binding expression (`buch !== undefined`) should itself be used as template guard:

```typescript
export declare class NgIf {
  static ngTemplateGuard_ngIf: 'verbatim'; // Just an example
}
```
This may then result in the following type-check code, which enables narrowing of `buch` to not include `undefined` in the if-block.
```typescript
if (buch !== undefined) {
  // ...
}
```

**Comment by JoostK** [comment]
The type narrowing limitation has been addressed in #30248 and accurate template source mapping was implemented in #30181 so this issue has been resolved.
